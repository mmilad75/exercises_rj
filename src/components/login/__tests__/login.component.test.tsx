import { render, screen } from "@testing-library/react";
import userEvent from "@testing-library/user-event";
import Login from "../login.component";

describe("Login component", () => {
    it("should render correctly", () => {
        const { asFragment } = render(<Login />);
        expect(asFragment()).toMatchSnapshot();
    });

    it("should allow user to type username", () => {
        render(<Login />);
        const testEmail = "mmilad@gmail.com";

        userEvent.type(
            screen.getByPlaceholderText("your-email@email.com"),
            testEmail
        );

        expect(screen.getByPlaceholderText("your-email@email.com")).toHaveValue(
            testEmail
        );
    });

    it("should allow user to type password", () => {
        render(<Login />);
        const testPassword = "1234$";

        userEvent.type(
            screen.getByPlaceholderText("Enter your password"),
            testPassword
        );

        expect(screen.getByPlaceholderText("Enter your password")).toHaveValue(
            testPassword
        );
    });

    it("should show alert on form submit", () => {
        render(<Login />);
        const mockAlert = jest.fn();
        jest.spyOn(window, "alert").mockImplementation(mockAlert);
        window.HTMLFormElement.prototype.submit = () => {};

        const testEmail = "mmilad@gmail.com";
        const testPassword = "1234$";

        userEvent.type(
            screen.getByPlaceholderText("your-email@email.com"),
            testEmail
        );
        userEvent.type(
            screen.getByPlaceholderText("Enter your password"),
            testPassword
        );
        userEvent.click(screen.getByRole("button"));

        expect(mockAlert).toBeCalledWith(
            `Login submitted: ${testEmail}, ${testPassword}`
        );
    });

    it("should show help alert after 5 seconds", () => {
        jest.useFakeTimers();

        render(<Login />);
        const mockAlert = jest.fn();
        jest.spyOn(window, "alert").mockImplementation(mockAlert);

        jest.advanceTimersByTime(6000);

        expect(mockAlert).toBeCalledWith("Do you need help?");
    });
});
