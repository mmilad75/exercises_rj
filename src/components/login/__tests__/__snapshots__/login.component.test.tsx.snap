// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login component should render correctly 1`] = `
<DocumentFragment>
  <form
    style="text-align: center; max-width: 400px; margin: 0px auto;"
  >
    <h1>
      Login
    </h1>
    <div>
      <div>
        <input
          name="email"
          placeholder="your-email@email.com"
          style="width: 100%; padding: 1rem; margin-bottom: 1rem; border-radius: 8px; background: rgb(225, 232, 237);"
          type="text"
          value=""
        />
      </div>
      <div>
        <input
          name="password"
          placeholder="Enter your password"
          style="width: 100%; padding: 1rem; margin-bottom: 1rem; border-radius: 8px; background: rgb(225, 232, 237);"
          type="password"
          value=""
        />
      </div>
    </div>
    <button
      style="padding: 0.5rem 1rem; margin: 0px 0.5rem; border-radius: 4px; background: rgb(20, 23, 26); color: rgb(255, 255, 255); font-size: 1rem;"
      type="submit"
    >
      Login
    </button>
  </form>
</DocumentFragment>
`;
